Source/NSGridView.m

```diff
@@ -0,0 +1,1233 @@
+/* Implementation of class NSGridView
+   Copyright (C) 2020 Free Software Foundation, Inc.
+   
+   By: Gregory John Casamento
+   Date: 08-08-2020
+
+   This file is part of the GNUstep Library.
+   
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+   
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+   
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; if not, write to the Free
+   Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110 USA.
+*/
+
+#import <Foundation/NSArray.h>
+#import "AppKit/NSGridView.h"
```

Comment:
I do not see any drawRect: methods in this file, how is this supposed to work out?

Reply:
OK, this "works" because you add the cell views as subviews of the grid view in _refreshCells. This is a rather simplistic way of implementing this. That won't allow you to have cell merging and it makes padding and alignment of cells rather hard. Are you sure to go ahead with that implementation?

Reply:
How else can I have functional views within the NSGridView?   What alternate implementation do you have in mind?  It is worth mentioning that when you list the subviews of the NSGridView that they are the actual views in the cells.   I would guess that suggests that my implementation is similar to the one done by Apple.

Reply:
Seeing this views as subview of the NSGridView is a good argument. I am surprised they did it that way. I expected that all the magic would happen in the drawRect: method. This has a few consequences I need to think about. The first being that you need to handle view removal. At the moment all cell views ever added remain as subviews of the NSGridView.
The next is the frame calculation for the views. Most likely you should start from your _prototypeFrame position that into the row/column and then subtract all the different padding values according to the placement.

Reply:
Just to document this in the PR.  The test for this class is here.. [https://github.com/gcasa/NSGridView_test.git](https://github.com/gcasa/NSGridView_test.git)

2021-03-26 02:24:57.844359-0400 NSGridView_test[62461:8943425] NSGridView has 13 subviews.
2021-03-26 02:24:57.844463-0400 NSGridView_test[62461:8943425] NSGridViewSizeForContent = 0.000000
2021-03-26 02:24:57.844530-0400 NSGridView_test[62461:8943425] Nunber of cells in row 1 = 9
2021-03-26 02:24:57.844860-0400 NSGridView_test[62461:8943425] Subviews = (
    "<NSButton: 0x7f81bb707aa0>",
    "<NSButton: 0x7f81bb40a6e0>",
    "<NSTextField: 0x7f81bb40b590>",
    "<NSColorWell: 0x7f81bb708b50>",
    "<NSStepper: 0x7f81bb709c80>",
    "<NSSwitch: 0x7f81bb70a8f0>",
    "<NSLevelIndicator: 0x7f81bb70b6c0>",
    "<NSSlider: 0x7f81bb523020>",
    "<NSView: 0x7f81bb40e4d0>",
    "<NSButton: 0x7f81bb40e9b0>",
    "<NSPopUpButton: 0x7f81bb40fa30>",
    "<NSButton: 0x7f81bb720e30>",
    "<NSButton: 0x7f81bc908d80>"
)